### PR TITLE
Consolidate and/or remove redundant or obsolete tests.

### PR DESCRIPTION
**Pull request by glennmatthews**
# Closes n/a (housekeeping)
# What's Changed

Targeting this to `next` due to potential impact downstream to apps.

- Deprecated the `FilterTestCases.NameOnlyFilterTestCase` and `FilterTestCases.NameSlugFilterTestCase` generic test classes. Apps should migrate to `FilterTestCases.FilterTestCase` with an appropriately defined list of `generic_filter_tests` instead.
- Fixed incorrect definition of `VirtualMachineFilterSet.cluster` filter. (it was an automatic single-value `ModelChoiceFilter` instead of a proper multiple-choice filter)
- Removed roughly 800 redundant and/or obsolete test cases. (In addition to the more specific changes below, replaced a lot of bespoke filterset tests with `generic_filter_tests` definitions)
    - Merged the `APIViewTestCases.NotesURLViewTestCase.test_notes_url_on_object` generic test function into `APIViewTestCases.GetObjectViewTestCase.test_get_object` generic test function to reduce redundant code and improve test speed.
    - Merged the `FilterTestCases.FilterTestCase.test_id` generic test function into `FilterTestCases.FilterTestCase.test_filters_generic` generic test function to reduce redundant code and improve test speed.
    - Removed the `FilterTestCases.FilterTestCase.test_q_filter_exists` generic test function as redundant.
    - Merged the `ViewTestCases.GetObjectViewTestCase.test_has_advanced_tab` generic test function into `ViewTestCases.GetObjectViewTestCase.test_get_object_with_permission` generic test function to reduce redundant code and improve test speed.
    - Removed the `ViewTestCases.CreateObjectViewTestCase.test_slug_autocreation` and `test_slug_not_modified` generic test functions as obsolete.
    - Merged the `ViewTestCases.ListObjectsViewTestCase.test_list_view_app_banner` generic test function into `ViewTestCases.ListObjectsViewTestCase.test_list_objects_with_permission` generic test function to reduce redundant code and improve test speed.
- Removed the previously deprecated `ViewTestCases.BulkImportObjectsViewTestCase` generic test class as obsolete.

Coverage results of running the full test suite before/after these changes show no increase in uncovered lines of code.

# TODO

- [x] Explanation of Change(s)
- [x] Added change log fragment(s) (for more information see [the documentation](https://docs.nautobot.com/projects/core/en/stable/development/#creating-changelog-fragments))
- n/a Attached Screenshots, Payload Example
- [x] Unit, Integration Tests
- n/a Documentation Updates (when adding/changing features)
- n/a Example App Updates (when adding/changing features)
- [x] Outline Remaining Work, Constraints from Design

### CODE REVIEW

File: nautobot/core/testing/api.py [hunk 1]
```diff
@@ -169,6 +169,10 @@ def test_get_object(self):
             if issubclass(self.model, extras_models.ChangeLoggedModel):
                 self.assertIn("created", response.data)
                 self.assertIn("last_updated", response.data)
+            if hasattr(self.model, "notes") and isinstance(instance1.notes, QuerySet):
+                self.assertIn("notes_url", response.data)
+                self.assertIn(f"{url}notes/", str(response.data["notes_url"]))
+                self.assertIn(instance1.get_notes_url(api=True), str(response.data["notes_url"]))
```
Comment (glennmatthews) [inline nautobot/core/testing/api.py:175]: Merged in from separate `test_notes_url_on_object` test case below.

File: nautobot/core/testing/filters.py [hunk 2]
```diff
@@ -81,19 +84,18 @@ class FilterTestCase(BaseFilterTestCase):
...
         #       ["filter2", "field2__name"],
         #   ]
-        generic_filter_tests = []
+        generic_filter_tests: ClassVar[Iterable]
+
+        def setUp(self):
+            for attr in ["queryset", "filterset", "generic_filter_tests"]:
+                if not hasattr(self, attr):
+                    raise NotImplementedError(f'{self} is missing a value for required attribute "{attr}"')
+            super().setUp()
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:93]: Changed these three attributes to abstract, and fail/abort test now if they're not defined by the subclass. I'd like to do the same for our other generic test classes that have required attributes.

File: nautobot/core/testing/filters.py [hunk 3]
```diff
@@ -81,19 +84,18 @@ class FilterTestCase(BaseFilterTestCase):
...
             """Helper method to return q filter."""
             return self.filterset.declared_filters["q"].filter_predicates
 
-        def test_id(self):
-            """Verify that the filterset supports filtering by id."""
-            params = {"id": self.queryset.values_list("pk", flat=True)[:2]}
-            filterset = self.filterset(params, self.queryset)
-            self.assertTrue(filterset.is_valid())
-            self.assertEqual(filterset.qs.count(), 2)
-
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:96]: Merged into `test_filters_generic` below.

File: nautobot/core/testing/filters.py [hunk 4]
```diff
@@ -219,10 +225,6 @@ def test_tags_filter(self):
             qs_result = self.queryset.filter(tags=tags[0]).filter(tags=tags[1]).distinct()
             self.assertQuerysetEqualAndNotEmpty(filterset_result, qs_result)
 
-        def test_q_filter_exists(self):
-            """Test the `q` filter exists on a filterset, does not validate the filter works as expected."""
-            self.assertIn("q", self.filterset.declared_filters)
-
```
Comment (glennmatthews) [inline nautobot/core/testing/filters.py:225]: Redundant with an existing check in `test_q_filter_valid`.

File: nautobot/core/testing/views.py [hunk 5]
```diff
@@ -170,48 +169,54 @@ def test_get_object_with_permission(self):
...
+                        else:
+                            self.assertIn(
+                                escape(str(instance.cf.get(custom_field.key) or "")), response_body, msg=response_body
+                            )
+
+            with self.subTest("Assert advanced tab is present"):
+                advanced_tab_href = f"{instance.get_absolute_url()}#advanced"
+
+                self.assertIn(advanced_tab_href, response_body)
+                self.assertIn("Advanced", response_body)
```
Comment (glennmatthews) [inline nautobot/core/testing/views.py:219]: Merged in from `test_has_advanced_tab` below.

File: nautobot/core/testing/views.py [hunk 6]
```diff
@@ -437,32 +422,6 @@ def test_extra_feature_form_fields_present(self):
...
-                expected_slug = self.slugify_function(getattr(obj, self.slug_source))
-                # Update slug source field str
-                filter_ = self.slug_source + "__exact"
-                self.model.objects.filter(**{filter_: self.slug_test_object}).update(
-                    **{self.slug_source: new_slug_source_value}
-                )
-
-                obj.refresh_from_db()
-                self.assertEqual(getattr(obj, self.slug_source), new_slug_source_value)
-                self.assertEqual(obj.slug, expected_slug)
```
Comment (glennmatthews) [inline nautobot/core/testing/views.py:464]: Removed as obsolete.

File: nautobot/core/testing/views.py [hunk 7]
```diff
@@ -876,17 +823,23 @@ def test_list_objects_with_permission(self):
...
+                self.assertNotIn(
+                    reverse("extras:job_run_by_class_path", kwargs={"class_path": "nautobot.core.jobs.ImportObjects"}),
+                    response_body,
+                )
+
+            if "example_app" in settings.PLUGINS:
+                with self.subTest("Assert example-app banner is present"):
+                    self.assertIn(
+                        f"<div>You are viewing a table of {self.model._meta.verbose_name_plural}</div>", response_body
+                    )
```
Comment (glennmatthews) [inline nautobot/core/testing/views.py:842]: Merged in from `test_list_view_app_banner` below.

File: nautobot/extras/tests/test_filters.py [hunk 8]
```diff
@@ -643,33 +542,17 @@ def test_associated_object_type(self):
         )
 
 
-class CustomFieldChoiceFilterSetTestCase(FilterTestCases.FilterTestCase):
```
Comment (HanlinMiao) [inline nautobot/extras/tests/test_filters.py:646]: Huh interesting we have two test cases for CustomFieldChoice there.

Also notice that we are naming things inconsistently throughout the test code. Some model filter tests are named *FilterSetTestCase and some are just *TestCase

File: nautobot/extras/tests/test_filters.py [hunk 9]
```diff
@@ -1669,6 +1368,12 @@ def test_destination_type(self):
 class RelationshipAssociationFilterSetTestCase(FilterTestCases.FilterTestCase):
     queryset = RelationshipAssociation.objects.all()
     filterset = RelationshipAssociationFilterSet
+    generic_filter_tests = (
+        ("destination_id",),
+        # ("relationship", "relationship__id"),  # TODO?
+        ("relationship", "relationship__key"),
```
Comment (HanlinMiao) [inline nautobot/extras/tests/test_filters.py:1374]: Is there a reason preventing us from changing `relationship` from `ModelMultipleChoiceFilter` to`NaturalKeyOrPKMultipleChoiceFilter`?

**Review by HanlinMiao — APPROVED** [review]

File: nautobot/extras/tests/test_filters.py (hunk 8 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_filters.py:646]: Yeah, that's another pass I'd like to do in "free" time, is standardize all of our test case naming patterns. :-)

File: nautobot/extras/tests/test_filters.py (hunk 9 above)
Comment (glennmatthews) [reply nautobot/extras/tests/test_filters.py:1374]: I don't think so, no, just wanted to keep the scope of this PR relatively constrained.

**Pull request merged**